base/lru_cache.hpp

```diff
@@ -0,0 +1,164 @@
+  /// All keys of |m_ageToKey| should be values of |m_ageToKey| and on the contrary
+  /// all keys of |m_ageToKey| should be values of |m_ageToKey|.
+  /// \note Ages should be unique for all keys.
+  class KeyAge
+  {
+    template <typename K, typename V> friend class LruCacheKeyAgeTest;
+  public:
+    /// \brief Increments |m_age| and insert key to |m_ageToKey| and |m_keyToAge|.
+    /// \note This method should be used only if there's no |key| in |m_ageToKey| and |m_keyToAge|.
+    void InsertKey(Key const & key)
+    {
+      ++m_age;
+      m_ageToKey[m_age] = key;
+      m_keyToAge[key] = m_age;
+    }
+
+    /// \brief Increments |m_age| and updates key age in |m_ageToKey| and |m_keyToAge|.
+    /// \note This method should be used only if there's |key| in |m_ageToKey| and |m_keyToAge|.
+    void UpdateAge(Key const & key)
+    {
+      ++m_age;
+      auto keyToAgeIt = m_keyToAge.find(key);
+      CHECK(keyToAgeIt != m_keyToAge.end(), ());
+      // Removing former age.
+      size_t const removed = m_ageToKey.erase(keyToAgeIt->second);
+      CHECK_EQUAL(removed, 1, ());
+      // Putting new age.
+      m_ageToKey[m_age] = key;
+      keyToAgeIt->second = m_age;
+    }
+
+    /// \returns Least recently used key without updating the age.
+    /// \note |m_ageToKey| and |m_keyToAge| shouldn't be empty.
+    Key const & GetLruKey() const
+    {
+      CHECK(!m_ageToKey.empty(), ());
+      // The smaller age the older item.
+      return m_ageToKey.cbegin()->second;
+    }
+
+    void RemoveLru()
+    {
+      Key const & lru = GetLruKey();
+      size_t const removed = m_keyToAge.erase(lru);
+      CHECK_EQUAL(removed, 1, ());
+      m_ageToKey.erase(m_ageToKey.begin());
+    }
+
+    /// \brief Checks for coherence class params.
+    /// \note It's a time consumption method and should be called for tests only.
+    bool IsValidForTesting() const
+    {
+      for (auto const & kv : m_ageToKey)
+      {
+        if (kv.first > m_age)
+          return false;
+        if (m_keyToAge.find(kv.second) == m_keyToAge.cend())
+          return false;
+      }
+      for (auto const & kv : m_keyToAge)
+      {
+        if (kv.second > m_age)
+          return false;
+        if (m_ageToKey.find(kv.second) == m_ageToKey.cend())
+          return false;
+      }
+      return true;
+    }
+
+    /// \returns true if |key| and its age are contained in |m_keyToAge| and |m_keyToAge|.
+    /// \note It's a time consumption method and should be called for tests only.
+    bool IsKeyValidForTesting(Key const & key) const
+    {
+      auto const keyToAgeId = m_keyToAge.find(key);
+      if (keyToAgeId == m_keyToAge.cend())
+        return false;
+      auto const ageToKeyIt = m_ageToKey.find(keyToAgeId->second);
+      if (ageToKeyIt == m_ageToKey.cend())
+        return false;
+      return key == ageToKeyIt->second;
+    }
+
+  private:
+    size_t m_age = 0;
+    std::map<size_t, Key> m_ageToKey;
```

Comment:
Strongly suggest you to change size_t to uint64_t. size_t may be 32 bits on some platforms (especially on devices), and, surprisingly, 2 ** 32 is not a big value :)

Reply:
Yes. I'll change that.